maitake-sync/src/wait_queue.rs

```diff
@@ -372,16 +391,52 @@ impl WaitQueue {
         // TODO(eliza): should this be a public API?
         #[must_use]
         pub(crate) fn new_woken() -> Self {
-            Self::new_with_state(State::Woken)
+            Self::make(State::Woken, Mutex::new(List::new()))
+        }
+    }
+}
+
+impl<Lock> WaitQueue<Lock>
+where
+    Lock: ScopedRawMutex + mutex_traits::ConstInit,
+{
+    loom_const_fn! {
+        /// Returns a new `WaitQueue`, using the provided [`ScopedRawMutex`]
+        /// implementation for wait-list synchronization.
+        ///
+        /// This constructor allows a `WaitQueue` to be constructed with any type that
+        /// implements [`ScopedRawMutex`] as the underlying raw blocking mutex
+        /// implementation. See [the documentation on overriding mutex
+        /// implementations](crate::blocking#overriding-mutex-implementations)
+        /// for more details.
+        #[must_use]
+        pub fn with_raw_mutex() -> Self {
```

Comment:
I don't think there's any way to make a non-default `WaitQueue` if `R: !ConstInit`, `::new()` creates a default, and `::with_raw_mutex()` requires `R: ConstInit`

Reply:
Yeah, I was planning to add a constructor that you just pass a `RawMutex` into, but I haven't actually written it yet because I wanted to figure out the `DefaultMutex` thing first.

Reply:
among other things, https://github.com/hawkw/mycelium/pull/482/commits/53e4b702493182a1499c38d79894b09a9fc4d3a9 changes all the `with_raw_mutex` constructors to take a caller-provided `RawMutex`, instead of using `ConstInit`. i may add additional constructors that use `ConstInit` later, but you can always do it yourself...